pyro/distributions/coalescent.py

```diff
@@ -0,0 +1,428 @@
+        represents a (batch of) rates that are piecwise constant in time.
+    """
+    arg_constraints = {"leaf_times": constraints.real,
+                       "rate_grid": constraints.positive}
+
+    def __init__(self, leaf_times, rate_grid, *, validate_args=None):
+        batch_shape = broadcast_shape(leaf_times.shape[:-1], rate_grid.shape[:-1])
+        event_shape = (leaf_times.size(-1) - 1,)
+        self.leaf_times = leaf_times
+        self.rate_grid = rate_grid
+        super().__init__(batch_shape, event_shape, validate_args=validate_args)
+
+    @constraints.dependent_property
+    def support(self):
+        return CoalescentTimesConstraint(self.leaf_times)
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(CoalescentTimesWithRate, _instance)
+        new.leaf_times = self.leaf_times
+        new.rate_grid = self.rate_grid
+        super(CoalescentTimesWithRate, new).__init__(
+            batch_shape, self.event_shape, validate_args=False)
+        new._validate_args = self.__dict__.get("_validate_args")
+        return new
+
+    def log_prob(self, value):
+        """
+        Computes likelihood as in equations 7-8 of [3].
+
+        This has time complexity ``O(T + S N log(N))`` where ``T`` is the
+        number of time steps, ``N`` is the number of leaves, and ``S =
+        sample_shape.numel()`` is the number of samples of ``value``.
+
+        This is differentiable wrt ``rate_grid`` but neither ``leaf_times`` nor
+        ``value = coal_times``.
+
+        :param torch.Tensor value: A tensor of coalescent times. These denote
+            sets of size ``leaf_times.size(-1) - 1`` along the trailing
+            dimension and should be sorted along that dimension.
+        :returns: Likelihood ``p(coal_times | leaf_times, rate_grid)``
+        :rtype: torch.Tensor
+        """
+        if self._validate_args:
+            self._validate_sample(value)
+        coal_times = value
+        phylogeny = _make_phylogeny(self.leaf_times, coal_times)
+
+        # Compute survival factors for closed intervals.
+        cumsum = self.rate_grid.cumsum(-1)
+        cumsum = torch.nn.functional.pad(cumsum, (1, 0), value=0)
+        integral = _interpolate_gather(cumsum, phylogeny.times[..., 1:])  # ignore the final lonely leaf
+        integral = integral[..., :-1] - integral[..., 1:]
+        integral = integral.clamp(min=torch.finfo(integral.dtype).tiny)  # avoid nan
+        log_prob = -(phylogeny.binomial[..., 1:-1] * integral).sum(-1)
+
+        # Compute density of coalescent events.
+        i = coal_times.floor().clamp(min=0).long()
+        rates = phylogeny.coal_binomial * _gather(self.rate_grid, -1, i)
+        log_prob = log_prob + safe_log(rates).sum(-1)
+
+        batch_shape = broadcast_shape(self.batch_shape, value.shape[:-1])
+        log_prob = log_prob.expand(batch_shape)
+        return log_prob
+
+
+class CoalescentRateLikelihood:
```

Comment:
Should this be a `nn.Module`?

Reply:
At the moment there is no reason to make this an `nn.Module`. It is a little more similar to a `Distribution`, but doesn't quite follow the distribution interface (indeed I've given it a `validate_args` kwarg like a distribution). Moreover in typical usage the parameters are all fixed rather than learned.

BTW this is a very natural object from the perspective of Funsor. Consider a funsor
```py
leaf_times = Tensor(...)
coal_times = Tensor(...)
rate_grid = Variable("rate_grid", reals(duration))
f = CoalescentTimesWithRate(leaf_times, rate_grid).log_prob(coal_times)
```
Then this likelihood represents roughly the funsor
```py
f(rate_grid=rate_grid["t"])
```
I look forward to the bright future when we won't need special classes like this 🚀